src/picostdlib/tusb.nim

```diff
@@ -64,71 +309,276 @@ type
+
+#  Clear the received FIFO
+#proc cdcReadFlush*(itf: uint8) {.importc: "tud_cdc_n_read_flush".}
+
+#  Get a byte from FIFO at the specified position without removing it
+#proc cdcPeek*(itf: uint8; ui8: ptr uint8): bool {.importc: "tud_cdc_n_peek".}
+
+#  Write bytes to TX FIFO, data may remain in the FIFO for a while
+proc cdcWrite*(itf: uint8; buffer: pointer; bufsize: uint32): uint32 {.importc: "tud_cdc_n_write".}
+
+#  Write a byte
+proc cdcWriteChar(itf: uint8; ch: char): uint32 {.inline, importc: "tud_cdc_n_write_char".}
+
+#  Write a null-terminated string
+#proc cdcWriteStr*(itf: uint8; str: cstring): uint32 {.inline, importc: "tud_cdc_n_write_str".}
+
+#  Force sending data if possible, return number of forced bytes
+proc cdcWriteFlush*(itf: uint8): uint32 {.importc: "tud_cdc_n_write_flush".}
+
+#  Return the number of bytes (characters) available for writing to TX FIFO buffer in a single n_write operation.
+#proc cdcWriteAvailable*(itf: uint8): uint32 {.importc: "tud_cdc_n_write_available".}
+
+#  Clear the transmit FIFO
+#proc cdcWriteClear*(itf: uint8): bool {.importc: "tud_cdc_n_write_clear".}
+{.pop}
+
+# Nim stream-like API for USB CDC (serial)
+
+type UsbSerialInterface* = distinct range[0'u8 .. 127'u8]
+## Used to represent the CDC interface number, as configured by `CFG_TUD_CDC`.
+## Eg. if `CFG_TUD_CDC` is `1`, then only `0.UsbSerialInterface` would be valid.
+## If `CFG_TUD_CDC` is `2`, then we could use `0.UsbSerialInterface` and
+## `1.UsbSerialInterface`.
+
+proc available*(itf: UsbSerialInterface): uint32 {.inline.} =
+  ## Get the number of bytes available for reading
+  cdcAvailable(itf.uint8)
+
+proc atEnd*(itf: UsbSerialInterface): bool {.inline.} =
+  ## Returns `true` if there is currently no data available for reading.
+  cdcAvailable(itf.uint8) == 0
+
+proc connected*(itf: UsbSerialInterface): bool {.inline.} =
+  ## Check if terminal is connected to this port
+  ##
+  ## Note: this actually checks for DTR bit, which is up to the host
+  ## application to set.
+  cdcConnected(itf.uint8)
+
+proc readBytes*(itf: UsbSerialInterface, n: Positive = 1): seq[uint8] {.inline.} =
+  ## Read up to `n` bytes from interface.
+  ##
+  ## Note: the length of the returned seq may be lower if fewer bytes were
+  ## actually read.
+  result = newSeq[uint8](n)
+  let actualNumBytes = cdcRead(itf.uint8, result[0].unsafeAddr, n.uint32)
+  result.setLen(actualNumBytes)
+
+proc readString*(itf: UsbSerialInterface, n: Positive = 1): string {.inline.} =
+  ## Read a string of up to `n` characters from interface.
+  ##
+  ## Note: the length of the returned string may be lower if fewer bytes were
+  ## actually read.
+  result = newString(n)
+  let actualNumBytes = cdcRead(itf.uint8, result[0].unsafeAddr, n.uint32)
+  result.setLen(actualNumBytes)
+
+template writeData(itf: uint8, data: untyped) =
+  var
+    i = 0
+    remain = data.len
+  while remain > 0 and cdcConnected(itf):
+    let buffer = data[i].unsafeAddr
```

Comment:
`data` is `untyped` but you then index `data`, should this be an explicit type?

Reply:
yeah, templates/macros are not my strong suit, ha. I wrote this one to reuse the code for the two `write` calls that can either take a string or byte array as `data`.

I think I could forego the template and make a generic `write` function instead, using a `T: string or openArray[uint8]` typeclass.

Reply:
As it's not exported it's probably fine though I guess probably should do
```nim
let cachedData = data
let buffer = cachedData[i].unsafeAddr
```

Reply:
Why make a copy? Especially on a micro I'd tend to avoid unecessary copies. The function is specifically written to not return until all data has been copied to tinyUSB's internal transmit FIFO, so no risk of trying to get data after the proc exits or anything.

Reply:
If `data` is a procedure call you're going to call it multiple times, though again possibly not an issue if only an internal tool.